pkg/tasks/update_template_content.go

```diff
@@ -0,0 +1,655 @@
+	if err != nil {
+		return nil, err
+	}
+
+	distResp, err := t.pulpClient.PollTask(t.ctx, *distTask)
+	if err != nil {
+		return nil, err
+	}
+
+	return distResp, nil
+}
+
+func (t *UpdateTemplateContent) createOrUpdateDistribution(distHref, distName, distPath, publicationHref string) error {
+	resp, err := t.pulpClient.FindDistributionByPath(t.ctx, distPath)
+	if err != nil {
+		return err
+	}
+
+	if resp == nil {
+		_, err := t.createDistributionWithContentGuard(publicationHref, distName, distPath)
+		if err != nil {
+			return err
+		}
+	} else {
+		taskHref, err := t.pulpClient.UpdateRpmDistribution(t.ctx, distHref, publicationHref, distName, distPath)
+		if err != nil {
+			return err
+		}
+
+		_, err = t.pulpClient.PollTask(t.ctx, taskHref)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getDistPathAndName(repo api.RepositoryResponse, templateUUID string, snapshotUUID string) (distPath string, distName string, err error) {
+	if repo.OrgID == config.RedHatOrg {
+		path, err := getRHRepoContentPath(repo.URL)
+		if err != nil {
+			return "", "", err
+		}
+		distPath = fmt.Sprintf("templates/%v/%v", templateUUID, path)
+	} else {
+		distPath = fmt.Sprintf("templates/%v/%v", templateUUID, repo.UUID)
+	}
+
+	distName = templateUUID + "/" + snapshotUUID
+	return distPath, distName, nil
+}
+
+func getRHRepoContentPath(rawURL string) (string, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return "", err
+	}
+	return u.Path[1 : len(u.Path)-1], nil
+}
+
+// RunCandlepin creates an environment for the template and content sets for each repository.
+// Each content set's URL is the distribution URL created during RunPulp().
+// May promote or demote content from the environment, depending on if the repository is being added or removed from template.
+// If not created for the given org previously, this will also create a product and pool.
+func (t *UpdateTemplateContent) RunCandlepin() error {
+	var err error
+
+	err = t.cpClient.CreateProduct(t.ctx, t.ownerKey)
+	if err != nil {
+		return err
+	}
+
+	poolID, err := t.cpClient.CreatePool(t.ctx, t.ownerKey)
+	if err != nil {
+		return err
+	}
+	t.payload.PoolID = &poolID
+	err = t.updatePayload()
+	if err != nil {
+		return err
+	}
+
+	customContent, customContentIDs, rhContentIDs, err := t.getContentList()
+	if err != nil {
+		return err
+	}
+
+	for _, item := range customContent {
+		err = t.cpClient.CreateContent(t.ctx, t.ownerKey, item)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO we can use create content batch when the api spec is fixed
+	//err = c.client.CreateContentBatch(candlepin_client.DevelOrgKey, content)
+	//if err != nil {
+	//	return err
+	//}
+
+	err = t.cpClient.AddContentBatchToProduct(t.ctx, t.ownerKey, customContentIDs)
+	if err != nil {
+		return err
+	}
+
+	contentPath, err := t.pulpClient.GetContentPath(t.ctx)
+	if err != nil {
+		return err
+	}
+	prefix := contentPath + t.rhDomainName + "/templates/" + t.payload.TemplateUUID
```

Comment:
I noticed that for me this is causing a resulting prefix of:
```
http://pulp.content:8080//pulp/content/80a1acac/templates/fc113564-fbcd-4ef8-8a3b-183015ef2ca7
```

the double //s isn't causing a problem, but you might use url.Parse & path.join() to do this

Reply:
https://stackoverflow.com/questions/34668012/combine-url-paths-with-path-join

Reply:
that's actually coming from the GetContentPath() function. It's because the the origin and prefix are defined with a trailing and leading slash:

```
PULP_CONTENT_ORIGIN: "http://pulp.content:8080/"
PULP_CONTENT_PATH_PREFIX: "/pulp/content/"
```

Still, it should handle that properly, so I'll add some logic there.